Modifications to improve testability

wilddeej opened the pull request:
Added optional command-line argument -t/--testing to indicate a test-only run, which generates client-specific text output instead of posting to the respective client.  A value can also be provided to this argument, which will influence the default configuration file name (feedspora.yml) and published DB name (feedspora.db).
If running in this test mode, no connections to the clients are made, setting the stage for automated testing of all clients/feeds without the security risk associated with client API keys or other such information to be present in the config files.
Finally, an initial test structure for using this new mechanism has been established, which includes both client-specific testing (with sub-options basic and full) and feed-specific testing (with sub-options atom and rss).  A *NIX bash shell script is also provided to drive these new tests.

aurelg commented on src/feedspora/__main__.py:
```diff
@@ -28,26 +29,42 @@ def read_config_file(filename):
+
+    # root name of config and DB files, optionally modified by the --testing
+    # argument value (if present)
+    root_name = 'feedspora'
```
The ternary operator would be nice: `root_name = args.testing if args.testing else 'feedspora'` (or maybe 'feedspora' could be the default value for parameter `testing`?) What do you think?

aurelg commented on tests/run_tests.sh:
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
```
The [unofficial bash strict mode](http://redsymbol.net/articles/unofficial-bash-strict-mode/) would be safer, just in case:

```
    #!/bin/bash
    set -euo pipefail
    IFS=$'\n\t'
```

aurelg replied:
That would also imply replacing `$1` by `${1:-}` to avoid referencing unbound variables all over the shell script. You may want to check [shellcheck](https://www.shellcheck.net)?

aurelg commented on tests/run_tests.sh:
```diff
@@ -0,0 +1,45 @@
+    filename=$(basename -- "$config")
+    test="${filename%.*}"
+    echo "    $test..."
+    rm -f ${test_dir}/${test}.out ${test_dir}/${test}.log ${test_dir}/${test}.db
```
Could it be `rm -f "${test_dir}"/"${test}".{out,log,db}` instead?

aurelg commented on tests/run_tests.sh:
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+function execute_tests() {
+  test_dir=$1
+  for config in `ls ${test_dir}/*.yml`; do
```
Couldn't it be simplified as `for config in "${test_dir}"/*.yml; do`?

aurelg commented:
I added a few comments in your code, just in case you'd like to improve your skills (`shellcheck`, in particular, is super useful). You don't have to fix it before I merge it, but if you want, you're most welcome. Just tell me :-)

wilddeej commented:
For __main__.py, I've implemented the ternary operator as you suggested,
but it's important in other places for the default to be None.  The
parameter value is optional, and in that case is configured to be
'feedspora', so I think that handles all the conditions properly.  At
least, it seems to be doing the right thing in my testing of that new
functionality.

I've added the changes you suggested to the run_tests.sh script as well.
Some planned updates to this script will require working around the set -e,
but it makes sense to handle those cases specifically different.

On Mon, Nov 26, 2018 at 9:08 AM Aurélien Grosdidier <
notifications@github.com> wrote:

> *@aurelg* commented on this pull request.
> ------------------------------
>
> In tests/run_tests.sh
> <https://github.com/aurelg/feedspora/pull/22#discussion_r236317509>:
>
> > @@ -0,0 +1,45 @@
> +#!/bin/bash
>
> The unofficial bash strict mode
> <http://redsymbol.net/articles/unofficial-bash-strict-mode/> would be
> safer, just in case:
>
>     #!/bin/bash
>     set -euo pipefail
>     IFS=$'\n\t'
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/aurelg/feedspora/pull/22#pullrequestreview-178368218>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AXIWhgdwVp1v9uAAAHTJSAPzR1xPE8Pbks5uzBIYgaJpZM4Yw4Mc>
> .
>

wilddeej commented:
I do plan on some changes to the run_tests.sh script specifically - OK if I
delay the remaining suggested changes until then?  It is totally functional
in its current form, and does what is intended.

On Mon, Nov 26, 2018 at 9:59 AM Jim Howard <wilddeej@gmail.com> wrote:

> For __main__.py, I've implemented the ternary operator as you suggested,
> but it's important in other places for the default to be None.  The
> parameter value is optional, and in that case is configured to be
> 'feedspora', so I think that handles all the conditions properly.  At
> least, it seems to be doing the right thing in my testing of that new
> functionality.
>
> I've added the changes you suggested to the run_tests.sh script as well.
> Some planned updates to this script will require working around the set -e,
> but it makes sense to handle those cases specifically different.
>
> On Mon, Nov 26, 2018 at 9:08 AM Aurélien Grosdidier <
> notifications@github.com> wrote:
>
>> *@aurelg* commented on this pull request.
>> ------------------------------
>>
>> In tests/run_tests.sh
>> <https://github.com/aurelg/feedspora/pull/22#discussion_r236317509>:
>>
>> > @@ -0,0 +1,45 @@
>> +#!/bin/bash
>>
>> The unofficial bash strict mode
>> <http://redsymbol.net/articles/unofficial-bash-strict-mode/> would be
>> safer, just in case:
>>
>>     #!/bin/bash
>>     set -euo pipefail
>>     IFS=$'\n\t'
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/aurelg/feedspora/pull/22#pullrequestreview-178368218>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AXIWhgdwVp1v9uAAAHTJSAPzR1xPE8Pbks5uzBIYgaJpZM4Yw4Mc>
>> .
>>
>

wilddeej commented:
For other reasons, it may end up smarter to replace run_tests.sh with a
Python driver instead in the near future.  As long as I can figure out how
to keep that as a separate module and not becoming "bound in" with
feedspora itself, that is.

On Mon, Nov 26, 2018 at 10:05 AM Jim Howard <wilddeej@gmail.com> wrote:

> I do plan on some changes to the run_tests.sh script specifically - OK if
> I delay the remaining suggested changes until then?  It is totally
> functional in its current form, and does what is intended.
>
> On Mon, Nov 26, 2018 at 9:59 AM Jim Howard <wilddeej@gmail.com> wrote:
>
>> For __main__.py, I've implemented the ternary operator as you suggested,
>> but it's important in other places for the default to be None.  The
>> parameter value is optional, and in that case is configured to be
>> 'feedspora', so I think that handles all the conditions properly.  At
>> least, it seems to be doing the right thing in my testing of that new
>> functionality.
>>
>> I've added the changes you suggested to the run_tests.sh script as well.
>> Some planned updates to this script will require working around the set -e,
>> but it makes sense to handle those cases specifically different.
>>
>> On Mon, Nov 26, 2018 at 9:08 AM Aurélien Grosdidier <
>> notifications@github.com> wrote:
>>
>>> *@aurelg* commented on this pull request.
>>> ------------------------------
>>>
>>> In tests/run_tests.sh
>>> <https://github.com/aurelg/feedspora/pull/22#discussion_r236317509>:
>>>
>>> > @@ -0,0 +1,45 @@
>>> +#!/bin/bash
>>>
>>> The unofficial bash strict mode
>>> <http://redsymbol.net/articles/unofficial-bash-strict-mode/> would be
>>> safer, just in case:
>>>
>>>     #!/bin/bash
>>>     set -euo pipefail
>>>     IFS=$'\n\t'
>>>
>>> —
>>> You are receiving this because you authored the thread.
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/aurelg/feedspora/pull/22#pullrequestreview-178368218>,
>>> or mute the thread
>>> <https://github.com/notifications/unsubscribe-auth/AXIWhgdwVp1v9uAAAHTJSAPzR1xPE8Pbks5uzBIYgaJpZM4Yw4Mc>
>>> .
>>>
>>
